src/zeroD/FlowDeviceFactory.cpp

```diff
@@ -0,0 +1,44 @@
+//! @file FlowDeviceFactory.cpp
+
+// This file is part of Cantera. See License.txt in the top-level directory or
+// at https://cantera.org/license.txt for license and copyright information.
+
+#include "cantera/zeroD/FlowDeviceFactory.h"
+#include "cantera/zeroD/FlowDeviceClasses.h"
+
+using namespace std;
+namespace Cantera
+{
+
+FlowDeviceFactory* FlowDeviceFactory::s_factory = 0;
+std::mutex FlowDeviceFactory::flowDevice_mutex;
+
+FlowDeviceFactory::FlowDeviceFactory()
+{
+    reg("MassFlowController", []() { return new MassFlowController(); });
+    reg("PressureController", []() { return new PressureController(); });
+    reg("Valve", []() { return new Valve(); });
+    // reg("FlowDevice", []() { return new FlowDevice(); }); // unused?
```

Comment:
I don't think the base class can be used directly, can it?

Reply:
It did have a constructor call in `ctreactor.cpp`, which was puzzling (hence my comment)